### PR TITLE
env(rjy): add ising model env

### PR DESCRIPTION
**Pull request by nighood**
## Related Issue
#773 

## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: dizoo/ising_env/envs/ising_model_env.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import numpy as np
+from numpy import dtype
+import gym
+from ding.envs import BaseEnv, BaseEnvTimestep
+from ding.torch_utils import to_ndarray, to_list
+from ding.utils import ENV_REGISTRY
+
+# ising_model_dir = os.path.join(os.path.dirname(__file__), 'ising_model')
+# if ising_model_dir not in sys.path:
+#     sys.path.append(ising_model_dir)
```
Comment (puyuan1996) [inline dizoo/ising_env/envs/ising_model_env.py:14]: 删除无用注释

File: dizoo/ising_env/envs/ising_model_env.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+# from ising_model.multiagent.environment import IsingMultiAgentEnv
+# import ising_model
+from dizoo.ising_env.envs.ising_model.multiagent.environment import IsingMultiAgentEnv
+import dizoo.ising_env.envs.ising_model as ising_model_
+
+
+@ENV_REGISTRY.register('ising_model')
+class IsingModelEnv(BaseEnv):
```
Comment (puyuan1996) [inline dizoo/ising_env/envs/ising_model_env.py:22]: 写一下overview，介绍ising_model的简介，agent目标，MDP定义，参考论文链接等

**Review by puyuan1996 — CHANGES_REQUESTED** [review]

File: ding/rl_utils/td.py [hunk 3]
```diff
@@ -688,7 +690,7 @@ def q_nstep_td_error(
     if weight is None:
         weight = torch.ones_like(reward)
 
-    if len(action.shape) == 1:  # single agent case
+    if len(action.shape) == 1 or len(action.shape) < len(q.shape):  # single agent case
```
[the remark this answers is not in the document]
Comment (nighood) [reply ding/rl_utils/td.py:693]: we need to unsqueeze action and q to make them have the same shape
e.g. single agent case: action is [B, ] and q is [B, ]
e.g. multi agent case: action is [B, agent_num] and q is [B, agent_num, action_shape]

File: ding/model/template/q_learning.py [hunk 4]
```diff
@@ -26,16 +26,17 @@ class DQN(nn.Module):
...
+        obs_shape: Union[int, SequenceType],
+        action_shape: Union[int, SequenceType],
+        encoder_hidden_size_list: SequenceType = [128, 128, 64],
+        dueling: bool = True,
+        head_hidden_size: Optional[int] = None,
+        head_layer_num: int = 1,
+        activation: Optional[nn.Module] = nn.ReLU(),
+        norm_type: Optional[str] = None,
+        dropout: Optional[float] = None,
+        init_bias: Optional[float] = None,
```
Comment (PaParaZz1) [inline ding/model/template/q_learning.py:39]: You need to also update the comments of `Arguments`

File: ding/policy/common_utils.py [hunk 5]
```diff
@@ -61,8 +61,17 @@ def default_preprocess_learn(
...
-        data['reward'] = reward.permute(1, 0).contiguous()
+        # single agent reward: (batch_size, nstep) -> (nstep, batch_size)
+        # multi-agent reward: (batch_size, agent_dim, nstep) -> (nstep, batch_size, agent_dim)
+        # Assuming 'reward' is a PyTorch tensor with shape (batch_size, nstep) or (batch_size, agent_dim, nstep)
+        if reward.ndim == 2:
+            # For a 2D tensor, simply transpose it to get (nstep, batch_size)
+            data['reward'] = reward.transpose(0, 1).contiguous()
+        elif reward.ndim == 3:
+            # For a 3D tensor, move the last dimension to the front to get (nstep, batch_size, agent_dim)
+            data['reward'] = reward.transpose(0, 2).transpose(1, 2).contiguous()
```
Comment (PaParaZz1) [inline ding/policy/common_utils.py:72]: use `permute` rather than two `transpose` operations

File: ding/policy/common_utils.py [hunk 6]
```diff
@@ -61,8 +61,17 @@ def default_preprocess_learn(
...
+        else:
+            raise ValueError("The 'reward' tensor must be either 2D or 3D.")
```
Comment (PaParaZz1) [inline ding/policy/common_utils.py:74]: You should add the ndim information in the error information

File: dizoo/ising_env/config/ising_mfq_config.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    policy=dict(type='dqn'),
+)
+ising_mfq_create_config = EasyDict(ising_mfq_create_config)
+create_config = ising_mfq_create_config
+
+if __name__ == '__main__':
+    # or you can enter `ding -m serial -c ising_mfq_config.py -s 0`
+    from ding.entry import serial_pipeline
+    seed = 1
+    set_pkg_seed(seed)
```
Comment (PaParaZz1) [inline dizoo/ising_env/config/ising_mfq_config.py:68]: why call this function, which is already called inside `serial_pipeline`

**Review by PaParaZz1 — APPROVED** [review]

**Review by PaParaZz1 — APPROVED** [review]

**Review by PaParaZz1 — APPROVED** [review]

**Pull request merged**